src/lib/serializer.ts

```diff
@@ -0,0 +1,253 @@
+    [21, 'l'],
+    [22, 'm'],
+    [23, 'n'],
+    [24, 'o'],
+    [25, 'p'],
+    [26, 'q'],
+    [27, 'r'],
+    [28, 's'],
+    [29, 't'],
+    [30, 'u'],
+    [31, 'v'],
+    [32, 'w'],
+    [33, 'x'],
+    [34, 'y'],
+    [35, 'z'],
+    [36, 'A'],
+    [37, 'B'],
+    [38, 'C'],
+    [39, 'D'],
+    [40, 'E'],
+    [41, 'F'],
+    [42, 'G'],
+    [43, 'H'],
+    [44, 'I'],
+    [45, 'J'],
+    [46, 'K'],
+    [47, 'L'],
+    [48, 'M'],
+    [49, 'N'],
+    [50, 'O'],
+    [51, 'P'],
+    [52, 'Q'],
+    [53, 'R'],
+    [54, 'S'],
+    [55, 'T'],
+    [56, 'U'],
+    [57, 'V'],
+    [58, 'W'],
+    [59, 'X'],
+    [60, 'Y'],
+    [61, 'Z'],
+    [62, '_'],
+    [63, '-'],
+  ]);
+
+  private static characterUnMap = new Map([
+    ['0', 0],
+    ['1', 1],
+    ['2', 2],
+    ['3', 3],
+    ['4', 4],
+    ['5', 5],
+    ['6', 6],
+    ['7', 7],
+    ['8', 8],
+    ['9', 9],
+    ['a', 10],
+    ['b', 11],
+    ['c', 12],
+    ['d', 13],
+    ['e', 14],
+    ['f', 15],
+    ['g', 16],
+    ['h', 17],
+    ['i', 18],
+    ['j', 19],
+    ['k', 20],
+    ['l', 21],
+    ['m', 22],
+    ['n', 23],
+    ['o', 24],
+    ['p', 25],
+    ['q', 26],
+    ['r', 27],
+    ['s', 28],
+    ['t', 29],
+    ['u', 30],
+    ['v', 31],
+    ['w', 32],
+    ['x', 33],
+    ['y', 34],
+    ['z', 35],
+    ['A', 36],
+    ['B', 37],
+    ['C', 38],
+    ['D', 39],
+    ['E', 40],
+    ['F', 41],
+    ['G', 42],
+    ['H', 43],
+    ['I', 44],
+    ['J', 45],
+    ['K', 46],
+    ['L', 47],
+    ['M', 48],
+    ['N', 49],
+    ['O', 50],
+    ['P', 51],
+    ['Q', 52],
+    ['R', 53],
+    ['S', 54],
+    ['T', 55],
+    ['U', 56],
+    ['V', 57],
+    ['W', 58],
+    ['X', 59],
+    ['Y', 60],
+    ['Z', 61],
+    ['_', 62],
+    ['-', 63],
+  ]);
+
+  // 0b00 reserved as error state
+  private static wallEncodeMap = new Map([
+    ['line-empty', 0b11],
+    ['line-wall', 0b01],
+    ['line-half', 0b10],
+  ]);
+
+  private static bitsPerWall = 2;
+  private static wallsPerByte = 3;
+
+  // Extracts the ith wall from an encoding in the characterUnMap
+  private static extractWallAtIndex(wallEncoding: number, i: number) {
+    const extractor = (1 << Serializer.bitsPerWall) - 1; // 0b11
+    const shift = i * Serializer.bitsPerWall;
+    const wallRepresentation = (wallEncoding >> shift) & extractor;
+    return wallRepresentation;
+  }
+
+  // Extracts a list of walls from a given wall encoding
+  private static extractWalls(wallEncoding: number) {
+    return [...Array(Serializer.wallsPerByte)].map((_, i) => {
+      return Serializer.extractWallAtIndex(wallEncoding, i);
+    });
+  }
+
+  private static packWalls(walls: number[]) {
+    return walls.reduce((previous, current, i) => {
+      const shift = Serializer.bitsPerWall * i;
+      return previous | (current << shift);
+    });
+  }
+
+  private static serializeWallsArray(elements: (SquareType | WallType)[]) {
+    const binaryList = elements
+      .filter((element) => 'className' in element)
```

Comment:
I would use `instanceof` here. Is there a reason you don't filter out SquareType in `serializeLayoutWalls`? It's a bit confusing to have `serializeWallsArray` also accept SquareType

Reply:
that's a good point, I'll filter it out before serializeWallsArray